be/src/exec/tablet_sink.cpp

```diff
@@ -172,138 +195,226 @@ Status NodeChannel::open_wait() {
+        SCOPED_TIMER(_parent->_compress_timer);
+
+        // Try compressing data to _compression_scratch, swap if compressed data is smaller
+        int max_compressed_size = _compress_codec->max_compressed_len(uncompressed_size);
+
+        if (_compression_scratch.size() < max_compressed_size) {
+            _compression_scratch.resize(max_compressed_size);
         }
 
-        if (result.has_execution_time_us()) {
-            _add_batch_counter.add_batch_execution_time_us += result.execution_time_us();
-            _add_batch_counter.add_batch_wait_lock_time_us += result.wait_lock_time_us();
-            _add_batch_counter.add_batch_num++;
+        Slice compressed_slice{_compression_scratch.data(), _compression_scratch.size()};
+        _compress_codec->compress(dst->data(), &compressed_slice);
+        double compress_ratio = (static_cast<double>(uncompressed_size)) / compressed_slice.size;
+        if (LIKELY(compress_ratio > config::rpc_compress_ratio_threshold)) {
+            _compression_scratch.resize(compressed_slice.size);
+            dst->mutable_data()->swap(reinterpret_cast<std::string&>(_compression_scratch));
+            dst->set_compress_type(_compress_type);
         }
-    });
 
-    return status;
+        VLOG_ROW << "uncompressed size: " << uncompressed_size << ", compressed size: " << compressed_slice.size;
+    }
+
+    return Status::OK();
 }
 
-Status NodeChannel::add_chunk(vectorized::Chunk* chunk, const int64_t* tablet_ids, const uint32_t* indexes,
-                              uint32_t from, uint32_t size) {
-    // If add_row() when _eos_is_produced==true, there must be sth wrong, we can only mark this channel as failed.
-    if (_cancelled | _eos_is_produced) {
+Status NodeChannel::add_chunk(vectorized::Chunk* input, const int64_t* tablet_ids, const uint32_t* indexes,
+                              uint32_t from, uint32_t size, bool eos) {
+    if (_cancelled || _send_finished) {
         return _err_st;
     }
 
-    // We use OlapTableSink mem_tracker which has the same ancestor of _plan node,
-    // so in the ideal case, mem limit is a matter for _plan node.
-    // But there is still some unfinished things, we do mem limit here temporarily.
-    // _cancelled may be set by rpc callback, and it's possible that _cancelled might be set in any of the steps below.
-    // It's fine to do a fake add_row() and return OK, because we will check _cancelled in next add_row() or mark_close().
-    while (!_cancelled && ((_mem_tracker->any_limit_exceeded() && _pending_batches_num > 0) ||
-                           _pending_batches_num >= _max_pending_batches_num)) {
-        SCOPED_RAW_TIMER(&_mem_exceeded_block_ns);
-        SleepFor(MonoDelta::FromMilliseconds(10));
-    }
+    RETURN_IF_ERROR(_wait_one_prev_request());
 
-    if (_cur_chunk->columns().empty()) {
-        _cur_chunk = chunk->clone_empty_with_slot();
+    if (!eos) {
+        SCOPED_TIMER(_parent->_pack_chunk_timer);
```

Comment:
Where is the mem_tracker to tracker the memory consumption?

Reply:
automatic tracer by the memory hook, the previous implementation has a individual queue memory tracker but no need now.